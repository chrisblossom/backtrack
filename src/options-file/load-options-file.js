/* @flow */

import cosmiconfig from 'cosmiconfig';
import { rootPath } from '../config/paths';
import type { Preset } from '../types.js';

function transform(result) {
    /**
     * Handle missing config file
     *
     * TODO: ask to create default backtrack config file with found presets in package.json
     */
    if (result === null) {
        throw new Error('backtrack config not found');
    }

    let config = result.config;

    /**
     * Handle ES Modules
     */
    if (typeof config === 'object' && config.__esModule) {
        if (config.default) {
            config = config.default;
        } else {
            throw new Error(
                `${result.filepath} must use default export with ES Modules`,
            );
        }
    }

    return {
        ...result,
        config,
    };
}

const explorer = cosmiconfig('backtrack', {
    rcExtensions: true,
    sync: true,
    transform,
});

function loadOptionsFile(searchPath: string = rootPath): Preset {
    const { config } = explorer.load(searchPath);

    return config;
}

export { loadOptionsFile };
