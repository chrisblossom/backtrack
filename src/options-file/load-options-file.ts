import { cosmiconfigSync } from 'cosmiconfig';
import { ErrorWithProcessExitCode } from '../utils/error-with-process-exit-code';
import { rootPath } from '../config/paths';
import { Preset } from '../types';

interface ESModule {
	__esModule: true;
	default?: Preset;
}

function isESModule(obj: any): obj is ESModule {
	// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
	return obj.__esModule === true;
}

interface TransformResult {
	config: Preset | ESModule;
	filepath: string;
}

function transform(result: TransformResult | null) {
	/**
	 * Handle missing config file
	 *
	 * TODO: ask to create default backtrack config file with found presets in package.json
	 */
	if (result === null) {
		const message = 'backtrack config not found';
		const exitCode = 1;
		throw new ErrorWithProcessExitCode(message, exitCode);
	}

	let config = result.config;

	/**
	 * Handle ES Modules
	 */
	if (isESModule(config)) {
		if (config.default) {
			config = config.default;
		} else {
			const message = `${result.filepath} must use default export with ES Modules`;
			const exitCode = 1;
			throw new ErrorWithProcessExitCode(message, exitCode);
		}
	}

	return {
		...result,
		config,
	};
}

const explorer = cosmiconfigSync('backtrack', {
	stopDir: rootPath,
	transform,
});

function loadOptionsFile(searchPath: string = rootPath): Preset {
	const result = explorer.search(searchPath);

	const config: Preset = result?.config;

	return config;
}

export { loadOptionsFile };
