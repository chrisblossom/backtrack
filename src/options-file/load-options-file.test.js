/* @flow */

import path from 'path';

const loadOptionsFile = (dir) =>
    require('./load-options-file').loadOptionsFile(dir);

describe('options', () => {
    const cwd = process.cwd();

    afterEach(() => {
        process.chdir(cwd);
    });

    test('handles module.exports', () => {
        const dir = path.resolve(__dirname, '__sandbox__/module-exports/');
        process.chdir(dir);

        const opts = loadOptionsFile();

        expect(opts).toMatchSnapshot();
    });

    test('handles module.exports - .backtrackrc.js', () => {
        const dir = path.resolve(__dirname, '__sandbox__/module-exports-rc/');
        process.chdir(dir);

        const opts = loadOptionsFile();

        expect(opts).toMatchSnapshot();
    });

    test('loads ES Modules config', () => {
        const dir = path.resolve(__dirname, '__sandbox__/es-modules/');
        process.chdir(dir);

        const opts = loadOptionsFile();

        expect(opts).toMatchSnapshot();
    });

    test('throws on missing config', () => {
        const dir = path.resolve(__dirname, '__sandbox__/missing/');
        process.chdir(dir);

        try {
            expect.hasAssertions();
            loadOptionsFile();
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('throws if ES Modules without default', () => {
        const dir = path.resolve(
            __dirname,
            '__sandbox__/es-modules-no-default/',
        );
        process.chdir(dir);

        try {
            expect.hasAssertions();
            loadOptionsFile();
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('handles custom searchPath', () => {
        const dir = path.resolve(__dirname, '__sandbox__/module-exports/');

        const opts = loadOptionsFile(dir);

        expect(opts).toMatchSnapshot();
    });
});
