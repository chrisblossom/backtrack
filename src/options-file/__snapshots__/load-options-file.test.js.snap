// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options handles custom searchPath 1`] = `
Object {
  "presets": Array [
    "module_exports",
  ],
}
`;

exports[`options handles module.exports - .backtrackrc.js 1`] = `
Object {
  "presets": Array [
    "module_exports_rc",
  ],
}
`;

exports[`options handles module.exports 1`] = `
Object {
  "presets": Array [
    "module_exports",
  ],
}
`;

exports[`options loads ES Modules config 1`] = `
Object {
  "presets": Array [
    "es_modules",
  ],
}
`;

exports[`options throws if ES Modules without default 1`] = `[Error: <PROJECT_ROOT>/backtrack.config.js must use default export with ES Modules]`;

exports[`options throws on missing config 1`] = `[Error: backtrack config not found]`;
